qubes/storage/zfs_encrypted.py

```diff
@@ -0,0 +1,502 @@
+                        unloaded'.format(self.zfs_ns))
+                    try:
+                        libzfs_core.lzc_unload_key(self.zfs_ns.encode())
+                        self.log.warning("UNLOADED key for {}.".format(
+                            self.zfs_ns))
+                    except libzfs_core.exceptions.EncryptionKeyNotLoaded:
+                        pass
+                    except libzfs_core.exceptions.ZFSError as e:
+                        self.log.warning(
+                            "key unloading failed for {}: {}".format(
+                                self.zfs_ns, e
+                            )
+                        )
+                        # try again:
+                        countdown = 10
+
+    async def ensure_key_is_loaded(self):
+        keystatus = qzfs.run_command(
+            ["zfs", "list", "-H",
+             "-o", 'keystatus', self.zfs_ns]
+        )
+        self.log.warning("track volume start keystatus {}".format(keystatus))
+        if keystatus.strip() == b'unavailable':
+            await self.ask_password(["zfs", "load-key", self.zfs_ns])
+            # TODO ideally I guess here we would wait for udevd to kick in...
+            await asyncio.sleep(1)
+        if keystatus.strip() == b'-':
+            self.log.warning("zfs track volume start err why is keystatus '-' ?")
+
+    @locked
+    async def track_volume_start(self, vm):
+        """
+        Register a volume (not a VM!) as used, for the purpose
+         of unloading the encryption key after a timeout
+         when no volumes are in use.
+        It will also prompt the user for their passphrase when trying
+         to start a volume in a group whose key is currently not loaded.
+        """
+        self.used_volumes.add(vm)
+        self.log.warning('track_volume_start add {} => {}'.format(
+                         vm, self.used_volumes))
+        await self.ensure_key_is_loaded()
+
+    @locked
+    async def track_volume_stop(self, vm):
+        """Register a volume (not a VM!) as NOT used anymore,
+           for the purpose of unloading the encryption key after a timeout
+           when no volumes are in use.
+        """
+        self.used_volumes.discard(vm)
+        self.used_volumes_last_empty = time.clock_gettime(
+            time.CLOCK_MONOTONIC_RAW)
+
+    def init_volume(self, appvm, volume_config):
+        vm_name = appvm.name
+        if not hasattr(volume_config, 'zfs_ns'):
+            volume_config["zfs_ns"] = self.zfs_ns
+        volume_config["pool"] = self
+        return ZFSQEncryptedVolume(
+            vm_name=vm_name,
+            encrypted_pool=self,
+            **volume_config
+        )
+
+    def included_in(self, app):
+        """
+        Returns the parent pool if found, otherwise raises an AssertionError.
+        This function also moonlights as our method to retrieve a handle for
+        'app' which we record and re-use when asking for passwords.
```

Comment:
As noted elsewhere, there is a better way: you can get it via `Pool.init_volume(vm)` -> `vm.app`, which is guaranteed to be called before any `volume.start()` - contrary to this function.